Support benchmark for polling ingestion

xuxiong1 opened the pull request:
### Description
This PR supports polling ingestion benchmarking, which reads from the workload and produces the message to a stream.
- Added a new operation-type: `ProduceStreamMessage`
- The new `ProduceStreamMessage(Runner)` reuses the existing `BulkIndexParamSource` to read the workload
- Added a `MessageProducerFactory` which creates the stream producer.
- Currently only support run in benchmark-only mode.
- Added shard-stats telemetry device endpoint

### Issues Resolved
related to issue: https://github.com/opensearch-project/OpenSearch/issues/17086
related to PR: https://github.com/opensearch-project/opensearch-benchmark-workloads/pull/578

### Testing
- [x] New functionality includes testing

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

xuxiong1 commented:
Local E2E test works:
```
 % opeopensearch-benchmark execute-test --workload-repository=default --workload-revision=982647ca64ceb45d9249c1ffa8e5ee06614fc6f3 --pipeline=benchmark-only --workload=geonames --target-host=127.0.0.1:9200 --workload-params '{"ingest_percentage":"1","bulk_size":"1000","number_of_shards":"1","number_of_replicas":"0","ingestion_source_type": "kafka","ingestion_topic":"test","ingestion_bootstrap_servers":"localhost:34803"}' --test-procedure=ingestion-index-only --kill-running-processes

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] [Test Execution ID]: 418f121d-a892-450d-a26a-1976a3d52689
[INFO] Executing test with workload [geonames], test_procedure [ingestion-index-only] and provision_config_instance ['external'] with version [3.0.0-SNAPSHOT].

Running delete-index                                                           [100% done]
Running create-index                                                           [100% done]
Running polling-ingest                                                         [100% done]

------------------------------------------------------
    _______             __   _____
   / ____(_)___  ____ _/ /  / ___/_________  ________
  / /_  / / __ \/ __ `/ /   \__ \/ ___/ __ \/ ___/ _ \
 / __/ / / / / / /_/ / /   ___/ / /__/ /_/ / /  /  __/
/_/   /_/_/ /_/\__,_/_/   /____/\___/\____/_/   \___/
------------------------------------------------------
            
|                                                         Metric |           Task |     Value |   Unit |
|---------------------------------------------------------------:|---------------:|----------:|-------:|
|                     Cumulative indexing time of primary shards |                |         0 |    min |
|             Min cumulative indexing time across primary shards |                |         0 |    min |
|          Median cumulative indexing time across primary shards |                |         0 |    min |
|             Max cumulative indexing time across primary shards |                |         0 |    min |
|            Cumulative indexing throttle time of primary shards |                |         0 |    min |
|    Min cumulative indexing throttle time across primary shards |                |         0 |    min |
| Median cumulative indexing throttle time across primary shards |                |         0 |    min |
|    Max cumulative indexing throttle time across primary shards |                |         0 |    min |
|                        Cumulative merge time of primary shards |                | 0.0203667 |    min |
|                       Cumulative merge count of primary shards |                |         2 |        |
|                Min cumulative merge time across primary shards |                | 0.0203667 |    min |
|             Median cumulative merge time across primary shards |                | 0.0203667 |    min |
|                Max cumulative merge time across primary shards |                | 0.0203667 |    min |
|               Cumulative merge throttle time of primary shards |                |         0 |    min |
|       Min cumulative merge throttle time across primary shards |                |         0 |    min |
|    Median cumulative merge throttle time across primary shards |                |         0 |    min |
|       Max cumulative merge throttle time across primary shards |                |         0 |    min |
|                      Cumulative refresh time of primary shards |                |   0.06665 |    min |
|                     Cumulative refresh count of primary shards |                |        23 |        |
|              Min cumulative refresh time across primary shards |                |   0.06665 |    min |
|           Median cumulative refresh time across primary shards |                |   0.06665 |    min |
|              Max cumulative refresh time across primary shards |                |   0.06665 |    min |
|                        Cumulative flush time of primary shards |                |         0 |    min |
|                       Cumulative flush count of primary shards |                |         0 |        |
|                Min cumulative flush time across primary shards |                |         0 |    min |
|             Median cumulative flush time across primary shards |                |         0 |    min |
|                Max cumulative flush time across primary shards |                |         0 |    min |
|                                        Total Young Gen GC time |                |     0.097 |      s |
|                                       Total Young Gen GC count |                |        19 |        |
|                                          Total Old Gen GC time |                |         0 |      s |
|                                         Total Old Gen GC count |                |         0 |        |
|                                                     Store size |                |  0.036993 |     GB |
|                                                  Translog size |                |         0 |     GB |
|                                         Heap used for segments |                |         0 |     MB |
|                                       Heap used for doc values |                |         0 |     MB |
|                                            Heap used for terms |                |         0 |     MB |
|                                            Heap used for norms |                |         0 |     MB |
|                                           Heap used for points |                |         0 |     MB |
|                                    Heap used for stored fields |                |         0 |     MB |
|                                                  Segment count |                |         3 |        |
|                                                 Min Throughput | polling-ingest |   1272.64 |  ops/s |
|                                                Mean Throughput | polling-ingest |   1368.68 |  ops/s |
|                                              Median Throughput | polling-ingest |   1372.62 |  ops/s |
|                                                 Max Throughput | polling-ingest |    1383.8 |  ops/s |
|                                        50th percentile latency | polling-ingest |   723.511 |     ms |
|                                        90th percentile latency | polling-ingest |   784.264 |     ms |
|                                        99th percentile latency | polling-ingest |   810.152 |     ms |
|                                       100th percentile latency | polling-ingest |   818.254 |     ms |
|                                   50th percentile service time | polling-ingest |   723.511 |     ms |
|                                   90th percentile service time | polling-ingest |   784.264 |     ms |
|                                   99th percentile service time | polling-ingest |   810.152 |     ms |
|                                  100th percentile service time | polling-ingest |   818.254 |     ms |
|                                                     error rate | polling-ingest |         0 |      % |


---------------------------------
[INFO] SUCCESS (took 111 seconds)
---------------------------------
%            
```

yupeng9 commented on tests/client_test.py:
```diff
@@ -467,3 +468,101 @@ def test_ssl_error(self, opensearch):
                                     expected_regex="Could not connect to cluster via https. Is this an https endpoint?"):
             client.wait_for_rest_layer(opensearch, max_attempts=3)
+
+class KafkaMessageProducerTests(TestCase):
```
does this require a kafka container turnup?

xuxiong1 replied:
Nope, we mocked it `@mock.patch("aiokafka.AIOKafkaProducer")`

xuxiong1 commented on osbenchmark/worker_coordinator/runner.py:
```diff
@@ -2832,3 +2833,147 @@ async def __call__(self, opensearch, params):
+            level="shards"
+        )
+
+        # Flatten the nested data structure to simplify processing
```
in push-based benchmark, it gets the processed items from the bulk request response, not from the node stats. Here we are parsing the node stats json returned.

xuxiong1 commented on osbenchmark/client.py:
```diff
@@ -430,3 +430,64 @@ def wait_for_rest_layer(opensearch, max_attempts=40):
+        else:
+            raise ValueError(f"Unsupported ingestion source type: {producer_type}")
+
+class KafkaMessageProducer:
```
yeah make sense.

yupeng9 commented:
LGTM

rishabh6788 commented on osbenchmark/worker_coordinator/runner.py:
```diff
@@ -2832,3 +2833,147 @@ async def __call__(self, opensearch, params):
+        total_processed_this_iteration = 0
+
+        # Get node stats with level=shards to access polling_ingest_stats
+        stats_response = await opensearch.nodes.stats(
```
My only concern is this part where we are running node-stats api as part of a regular run. We support node-stats api as part of `--telemetry` devices, which when enabled will independently run node-stats api, every 1s, and ingest data into `benchmark-metrics-*` index, See https://github.com/opensearch-project/opensearch-benchmark/blob/main/osbenchmark/telemetry.py#L652. 
https://opensearch.org/docs/latest/benchmark/reference/telemetry/#node-stats

Do you need this api to print data as part of final results?

rishabh6788 replied:
I'm good with the rest.

xuxiong1 replied:
Yeah, that's a good idea, retrieving the node stats in a separate run could be better, I'll make the change accordingly, thanks!

xuxiong1 replied:
Hi @rishabh6788, 
I found in the wiki page there are [shard-stats](https://opensearch.org/docs/latest/benchmark/reference/telemetry/#shard-stats) telemetry device which I could use, but while checking the code base, seems like this shard-level device is not there, do we know the reason why it's removed?
```
 % opensearch-benchmark list telemetry

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

Available telemetry devices:

Command                     Name                        Description
--------------------------  --------------------------  --------------------------------------------------------------------------------------------------------------------------------
jit                         JIT Compiler Profiler       Enables JIT compiler logs.
gc                          GC log                      Enables GC logs.
jfr                         Flight Recorder             Enables Java Flight Recorder (requires an Oracle JDK or OpenJDK 11+)
heapdump                    Heap Dump                   Captures a heap dump.
node-stats                  Node Stats                  Regularly samples node stats
recovery-stats              Recovery Stats              Regularly samples shard recovery stats
ccr-stats                   CCR Stats                   Regularly samples Cross Cluster Replication (CCR) leader and follower(s) checkpoint at index leveland calculates replication lag
segment-stats               Segment Stats               Determines segment stats at the end of the benchmark.
transform-stats             Transform Stats             Regularly samples transform stats
searchable-snapshots-stats  Searchable Snapshots Stats  Regularly samples searchable snapshots stats
segment-replication-stats   Segment Replication Stats   Regularly samples segment replication stats
```

rishabh6788 replied:
hmm, that'w weird, I do not see shardstats in the code. 
You can add it as a new telemetry device if that helps?
You can refer the apache licensed rally implementation, https://github.com/elastic/rally/blob/master/esrally/telemetry.py#L661. 
Modify it as per your need.

IanHoang commented:
@xuxiong1 Unittests are failing right now due to the following error:
```
ERROR: Could not find a version that satisfies the requirement aiokafka>=0.12.0 (from opensearch-benchmark) (from versions: 0.0.1, 0.1.0, 0.1.1, 0.1.2, 0.1.3, 0.1.4, 0.2.0, 0.2.1, 0.2.2, 0.2.3, 0.3.0, 0.3.1, 0.4.0, 0.4.1, 0.4.2.dev0, 0.4.2, 0.4.3, 0.5.0, 0.5.1, 0.5.2, 0.6.0, 0.7.0, 0.7.1, 0.7.2, 0.8.0, 0.8.1, 0.9.0rc0, 0.9.0rc1, 0.9.0, 0.10.0a0, 0.10.0, 0.11.0)
ERROR: No matching distribution found for aiokafka>=0.12.0
```

Although aiokafka `0.12.0` is available on PyPi, OSB does not recognize it as an available version. Could you verify that aiokafka version `0.12.0` was installed in the environment where you performed the E2E test? 

For now, we can quickly use `aiokafka>=0.11.0` in `setup.py`.

xuxiong1 commented:
local E2E test with telemetry enabled:
```
 % opensearch-benchmark execute-test \
  --workload-repository=default \
  --workload-revision=51899e7b8b21f23b8050b5ad363c65e96d17d757 \
  --pipeline=benchmark-only \
  --workload=geonames \
  --target-host=127.0.0.1:9200 \
  --workload-params '{
    "ingest_percentage": "5",
    "bulk_size": "1000",
    "number_of_shards": "1",
    "number_of_replicas": "0",
    "ingestion_source_type": "kafka",
    "ingestion_topic": "test",
    "ingestion_bootstrap_servers": "localhost:44469",
    "ingestion_pointer_init_reset": "earliest"
  }' \
  --test-procedure=polling-ingest-index-only \
  --kill-running-processes \
  --telemetry shard-stats \
  --telemetry-params '{
    "shard-stats-sample-interval": 0.1
  }'
```
test output:
```
   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] [Test Execution ID]: 6b328b3e-16df-43cb-a952-674800333268
[INFO] Downloading workload data file: documents-2.json.bz2 (252.9 MB total size) [100.0%]
[INFO] Decompressing workload data from [/home/user/.osb/benchmarks/data/geonames/documents-2.json.bz2] to [/home/user/.osb/benchmarks/data/geonames/documents-2.json] (resulting size: [3.30] GB) ... [OK]
[INFO] Downloading workload data file: documents-2.json.offset (4.2 kB total size)[100.0%]
[INFO] Executing test with workload [geonames], test_procedure [polling-ingest-index-only] and provision_config_instance ['external'] with version [3.0.0-alpha1].

[WARNING] merges_total_time is 43043 ms indicating that the cluster is not in a defined clean state. Recorded index time metrics may be misleading.
[WARNING] merges_total_throttled_time is 19640 ms indicating that the cluster is not in a defined clean state. Recorded index time metrics may be misleading.
[WARNING] indexing_total_time is 3253 ms indicating that the cluster is not in a defined clean state. Recorded index time metrics may be misleading.
[WARNING] refresh_total_time is 15191 ms indicating that the cluster is not in a defined clean state. Recorded index time metrics may be misleading.
[WARNING] flush_total_time is 1914 ms indicating that the cluster is not in a defined clean state. Recorded index time metrics may be misleading.
Running delete-index                                                           [100% done]
Running create-index                                                           [100% done]
Running polling-ingest                                                         [100% done]
Running refresh-after-index                                                    [100% done]
Running force-merge                                                            [100% done]
Running refresh-after-force-merge                                              [100% done]
Running wait-until-merges-finish                                               [100% done]

------------------------------------------------------
    _______             __   _____
   / ____(_)___  ____ _/ /  / ___/_________  ________
  / /_  / / __ \/ __ `/ /   \__ \/ ___/ __ \/ ___/ _ \
 / __/ / / / / / /_/ / /   ___/ / /__/ /_/ / /  /  __/
/_/   /_/_/ /_/\__,_/_/   /____/\___/\____/_/   \___/
------------------------------------------------------
            
|                                                         Metric |                     Task |     Value |   Unit |
|---------------------------------------------------------------:|-------------------------:|----------:|-------:|
|                     Cumulative indexing time of primary shards |                          |     0.107 |    min |
|             Min cumulative indexing time across primary shards |                          |         0 |    min |
|          Median cumulative indexing time across primary shards |                          |         0 |    min |
|             Max cumulative indexing time across primary shards |                          |  0.105083 |    min |
|            Cumulative indexing throttle time of primary shards |                          |         0 |    min |
|    Min cumulative indexing throttle time across primary shards |                          |         0 |    min |
| Median cumulative indexing throttle time across primary shards |                          |         0 |    min |
|    Max cumulative indexing throttle time across primary shards |                          |         0 |    min |
|                        Cumulative merge time of primary shards |                          |   0.82095 |    min |
|                       Cumulative merge count of primary shards |                          |        26 |        |
|                Min cumulative merge time across primary shards |                          |         0 |    min |
|             Median cumulative merge time across primary shards |                          |         0 |    min |
|                Max cumulative merge time across primary shards |                          |  0.813483 |    min |
|               Cumulative merge throttle time of primary shards |                          |    0.3993 |    min |
|       Min cumulative merge throttle time across primary shards |                          |         0 |    min |
|    Median cumulative merge throttle time across primary shards |                          |         0 |    min |
|       Max cumulative merge throttle time across primary shards |                          |    0.3993 |    min |
|                      Cumulative refresh time of primary shards |                          |   0.32125 |    min |
|                     Cumulative refresh count of primary shards |                          |       298 |        |
|              Min cumulative refresh time across primary shards |                          |         0 |    min |
|           Median cumulative refresh time across primary shards |                          |         0 |    min |
|              Max cumulative refresh time across primary shards |                          |  0.301117 |    min |
|                        Cumulative flush time of primary shards |                          | 0.0204167 |    min |
|                       Cumulative flush count of primary shards |                          |        14 |        |
|                Min cumulative flush time across primary shards |                          |         0 |    min |
|             Median cumulative flush time across primary shards |                          |         0 |    min |
|                Max cumulative flush time across primary shards |                          | 0.0164667 |    min |
|                                        Total Young Gen GC time |                          |     0.861 |      s |
|                                       Total Young Gen GC count |                          |       240 |        |
|                                          Total Old Gen GC time |                          |         0 |      s |
|                                         Total Old Gen GC count |                          |         0 |        |
|                                                     Store size |                          |  0.966518 |     GB |
|                                                  Translog size |                          | 0.0864068 |     GB |
|                                         Heap used for segments |                          |         0 |     MB |
|                                       Heap used for doc values |                          |         0 |     MB |
|                                            Heap used for terms |                          |         0 |     MB |
|                                            Heap used for norms |                          |         0 |     MB |
|                                           Heap used for points |                          |         0 |     MB |
|                                    Heap used for stored fields |                          |         0 |     MB |
|                                                  Segment count |                          |        53 |        |
|                                                 Min Throughput |           polling-ingest |   1121.74 |  ops/s |
|                                                Mean Throughput |           polling-ingest |    1349.2 |  ops/s |
|                                              Median Throughput |           polling-ingest |    1355.8 |  ops/s |
|                                                 Max Throughput |           polling-ingest |   1366.79 |  ops/s |
|                                        50th percentile latency |           polling-ingest |   730.485 |     ms |
|                                        90th percentile latency |           polling-ingest |   784.434 |     ms |
|                                        99th percentile latency |           polling-ingest |   882.368 |     ms |
|                                       100th percentile latency |           polling-ingest |   1007.32 |     ms |
|                                   50th percentile service time |           polling-ingest |   730.485 |     ms |
|                                   90th percentile service time |           polling-ingest |   784.434 |     ms |
|                                   99th percentile service time |           polling-ingest |   882.368 |     ms |
|                                  100th percentile service time |           polling-ingest |   1007.32 |     ms |
|                                                     error rate |           polling-ingest |         0 |      % |
|                                                 Min Throughput | wait-until-merges-finish |     157.7 |  ops/s |
|                                                Mean Throughput | wait-until-merges-finish |     157.7 |  ops/s |
|                                              Median Throughput | wait-until-merges-finish |     157.7 |  ops/s |
|                                                 Max Throughput | wait-until-merges-finish |     157.7 |  ops/s |
|                                       100th percentile latency | wait-until-merges-finish |   5.99734 |     ms |
|                                  100th percentile service time | wait-until-merges-finish |   5.99734 |     ms |
|                                                     error rate | wait-until-merges-finish |         0 |      % |


---------------------------------
[INFO] SUCCESS (took 476 seconds)
---------------------------------
```

able to visualize the metrics on opensearch dashboard:
<img width="2554" alt="Screenshot 2025-03-22 at 8 06 47 PM" src="https://github.com/user-attachments/assets/dbc2bb7a-c5a7-42c9-ba02-b069e8d03e9c" />

Hi @rishabh6788 @IanHoang @yupeng9 updated the PR with telemetry enabled, please take a look.

rishabh6788 commented:
LGTM, please fix the lint errors.

xuxiong1 commented:
Hi @rishabh6788 , I updated the PR and resolved the lint errors, please take a look.
Also I moved `RequestContextHolder` to a new context.py file to resolve the cyclic dependency issue, as it's imported by both osbenchmark.client and osbenchmark.kafka_client.

rishabh6788 commented:
@IanHoang @gkamat I'm good, feel free to merge if no concerns.

xuxiong1 commented:
@IanHoang @gkamat gentle ping here in case you missed.
